R/input-switch.R

```diff
@@ -78,16 +78,48 @@ toggle_switch <- function(id, value = NULL, session = get_current_session()) {
 }
 
 # Keep this internal for now until we agree on the UX
-input_dark_mode_switch <- function(id, label, ..., width = "auto") {
-  res <- input_switch(id, label, ..., width = width)
-  # TODO: .navbar should probably do this for .shiny-input-container
-  res <- tagAppendAttributes(res, style = "margin-bottom: 0;")
-  tagAppendAttributes(
-    res, .cssSelector = "input",
-    onclick = "document.documentElement.setAttribute('data-bs-theme', this.checked ? 'dark' : 'light'); $(window).resize();"
+input_dark_mode_switch <- function(id, ..., mode = NULL) {
+  if (!is.null(mode)) {
+    mode <- rlang::arg_match(mode, c("light", "dark"))
+  }
+
+  if (any(!nzchar(rlang::names2(rlang::list2(...))))) {
+    abort("All arguments in `...` must be named.")
+  }
+
+  res <- web_component(
+    "bslib-dark-mode-switch",
+    id = id,
+    "theme-attribute" = "data-bs-theme",
+    "theme-value" = mode,
+    style = css(
+      "--text-1" = "var(--bs-emphasis-color)",
+      "--text-2" = "var(--bs-tertiary-color)",
+      # The vertical correction used in the dark mode component isn't quite
+      # right on Bootstrap pages. This next line overrides it and removes the
+      # vertical correction. But users can still set the CSS property manually
+      # in their a `style` argument passed in via `...` if they want.
+      "--vertical-correction" = " "
+    ),
+    ...
   )
+
+  res <- tag_require(res, version = 5, caller = "input_dark_mode_switch()")
+  as_fragment(res)
 }
 
+toggle_dark_mode_switch <- function(id, mode = NULL, session = get_current_session()) {
+  mode <- mode %||% "toggle"
+  mode <- rlang::arg_match(mode, c("light", "dark", "toggle"))
+
+  if (!rlang::is_string(id)) {
+    abort("`id` must be a single string containing the `id` of a switch input.")
+  }
+
+  session$sendInputMessage(id, list(method = "toggle", value = mode))
```

Comment:
Is `method` necessary? I think we could just use `value` here (especially if the function is changed to `update_dark_mode_switch`).

Reply:
No it's not strictly necessary, but it's consistent with other input bindings in bslib and leaves room for future input message methods